Downloading a big Multipart file on `resteasy-reactive` is throwing an `OutOfMemoryError` on Upstream

### Describe the bug

Quarkus version: upstream
Extension: `quarkus-rest-client-reactive-jackson`

When I try to download a ~2Gb multipart file I am getting the following error
```
18:23:06,870 INFO  [app] Caused by: java.lang.OutOfMemoryError: Required array length 2147483639 + 9 is too large
18:23:06,871 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.hugeLength(ArraysSupport.java:649)
18:23:06,872 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.newLength(ArraysSupport.java:642)
18:23:06,872 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:100)
18:23:06,873 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:130)
18:23:06,874 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.doWrite(FileBodyHandler.java:64)
18:23:06,874 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:56)
18:23:06,875 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:22)
18:23:06,876 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeEntity(MultipartMessageBodyWriter.java:189)
18:23:06,876 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writePart(MultipartMessageBodyWriter.java:130)
18:23:06,877 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.write(MultipartMessageBodyWriter.java:94)
18:23:06,878 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeMultiformData(MultipartMessageBodyWriter.java:68)
18:23:06,878 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeResponse(MultipartMessageBodyWriter.java:50)
18:23:06,879 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:220)
18:23:06,880 INFO  [app]        ... 12 more
18:23:06,880 INFO  [app] 18:23:05,789 Request failed: java.lang.RuntimeException: java.lang.OutOfMemoryError: Required array length 2147483639 + 9 is too large
18:23:06,881 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:258)
18:23:06,882 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:191)
18:23:06,882 INFO  [app]        at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriter.write(FixedEntityWriter.java:28)
18:23:06,883 INFO  [app]        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:34)
18:23:06,884 INFO  [app]        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:120)
18:23:06,884 INFO  [app]        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
18:23:06,885 INFO  [app]        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:564)
18:23:06,885 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
18:23:06,886 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
18:23:06,886 INFO  [app]        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
18:23:06,887 INFO  [app]        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
18:23:06,888 INFO  [app]        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
18:23:06,888 INFO  [app]        at java.base/java.lang.Thread.run(Thread.java:833)
18:23:06,889 INFO  [app] Caused by: java.lang.OutOfMemoryError: Required array length 2147483639 + 9 is too large
18:23:06,889 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.hugeLength(ArraysSupport.java:649)
18:23:06,890 INFO  [app]        at java.base/jdk.internal.util.ArraysSupport.newLength(ArraysSupport.java:642)
18:23:06,890 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:100)
18:23:06,890 INFO  [app]        at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:130)
18:23:06,891 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.doWrite(FileBodyHandler.java:64)
18:23:06,891 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:56)
18:23:06,891 INFO  [app]        at org.jboss.resteasy.reactive.common.providers.serialisers.FileBodyHandler.writeTo(FileBodyHandler.java:22)
18:23:06,892 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeEntity(MultipartMessageBodyWriter.java:189)
18:23:06,892 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writePart(MultipartMessageBodyWriter.java:130)
18:23:06,892 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.write(MultipartMessageBodyWriter.java:94)
18:23:06,893 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeMultiformData(MultipartMessageBodyWriter.java:68)
18:23:06,893 INFO  [app]        at org.jboss.resteasy.reactive.server.core.multipart.MultipartMessageBodyWriter.writeResponse(MultipartMessageBodyWriter.java:50)
18:23:06,893 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:220)
18:23:06,893 INFO  [app]        ... 12 more
```

Multipart pojo:
```
public class FileWrapper {
    @RestForm("file")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public File file;

    @RestForm("name")
    @PartType(MediaType.TEXT_PLAIN)
    public String name;
}

```

Rest API (that is failing)
```
    @GET
    @Produces(MediaType.MULTIPART_FORM_DATA)
    @Path("/download-multipart")
    public FileWrapper downloadMultipart() {
        FileWrapper wrapper = new FileWrapper();
        wrapper.file = file;
        wrapper.name = file.getName();
        return wrapper;
    }
```
**Note:** `File` is a 2Gb file. Quarkus 2.13.3.Final works as expected! 


### Expected behavior

No error 

### Actual behavior

outofMemory exception

### How to Reproduce?

`git clone git@github.com:quarkus-qe/quarkus-test-suite.git`
`cd http/rest-client-reactive`
`mvn clean verify -Dit.test=LargeFileHandlingIT#downloadMultipart` 

**Note:** double-check that `downloadMultipart` is not disabled. 

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "17.0.4" 2022-07-19 OpenJDK Runtime Environment Temurin-17.0.4+8 (build 17.0.4+8) OpenJDK 64-Bit Server VM Temurin-17.0.4+8 (build 17.0.4+8, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/independent-projects/resteasy-reactive/server/jsonb/src/main/java/org/jboss/resteasy/reactive/server/jsonb/JsonbMessageBodyWriter.java
+++ b/independent-projects/resteasy-reactive/server/jsonb/src/main/java/org/jboss/resteasy/reactive/server/jsonb/JsonbMessageBodyWriter.java
@@ -13,6 +13,7 @@ import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.MultivaluedMap;
 
 import org.jboss.resteasy.reactive.common.providers.serialisers.JsonMessageBodyWriterUtil;
+import org.jboss.resteasy.reactive.server.NoopCloseAndFlushOutputStream;
 import org.jboss.resteasy.reactive.server.StreamingOutputStream;
 import org.jboss.resteasy.reactive.server.spi.ServerMessageBodyWriter;
 import org.jboss.resteasy.reactive.server.spi.ServerRequestContext;
@@ -50,41 +51,5 @@ public class JsonbMessageBodyWriter extends ServerMessageBodyWriter.AllWriteable
         }
         // we don't use try-with-resources because that results in writing to the http output without the exception mapping coming into play
         originalStream.close();
-    }
-
-    /**
-     * This class is needed because Yasson doesn't give us a way to control if the output stream is going to be closed or not
-     */
-    private static class NoopCloseAndFlushOutputStream extends OutputStream {
-        private final OutputStream delegate;
-
-        public NoopCloseAndFlushOutputStream(OutputStream delegate) {
-            this.delegate = delegate;
-        }
-
-        @Override
-        public void flush() {
-
-        }
-
-        @Override
-        public void close() {
-
-        }
-
-        @Override
-        public void write(int b) throws IOException {
-            delegate.write(b);
-        }
-
-        @Override
-        public void write(byte[] b) throws IOException {
-            delegate.write(b);
-        }
-
-        @Override
-        public void write(byte[] b, int off, int len) throws IOException {
-            delegate.write(b, off, len);
-        }
     }
 }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/NoopCloseAndFlushOutputStream.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/NoopCloseAndFlushOutputStream.java
@@ -1,0 +1,41 @@
+package org.jboss.resteasy.reactive.server;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * This class is needed because some message writers don't give us a way to control if the output stream is going to be closed
+ * or not.
+ */
+public class NoopCloseAndFlushOutputStream extends OutputStream {
+    private final OutputStream delegate;
+
+    public NoopCloseAndFlushOutputStream(OutputStream delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void flush() {
+
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        delegate.write(b);
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+        delegate.write(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        delegate.write(b, off, len);
+    }
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultipartMessageBodyWriter.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultipartMessageBodyWriter.java
@@ -1,6 +1,5 @@
 package org.jboss.resteasy.reactive.server.core.multipart;
 
-import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.OutputStream;
@@ -24,6 +23,7 @@ import javax.ws.rs.ext.MessageBodyWriter;
 import org.jboss.resteasy.reactive.common.core.Serialisers;
 import org.jboss.resteasy.reactive.common.reflection.ReflectionBeanFactoryCreator;
 import org.jboss.resteasy.reactive.multipart.FileDownload;
+import org.jboss.resteasy.reactive.server.NoopCloseAndFlushOutputStream;
 import org.jboss.resteasy.reactive.server.core.CurrentRequestManager;
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.jboss.resteasy.reactive.server.core.ServerSerialisers;
@@ -184,11 +184,10 @@ public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWrite
         boolean wrote = false;
         for (MessageBodyWriter<Object> writer : writers) {
             if (writer.isWriteable(entityClass, entityType, Serialisers.NO_ANNOTATION, mediaType)) {
-                try (ByteArrayOutputStream writerOutput = new ByteArrayOutputStream()) {
+                try (NoopCloseAndFlushOutputStream writerOutput = new NoopCloseAndFlushOutputStream(os)) {
                     // FIXME: spec doesn't really say what headers we should use here
                     writer.writeTo(entity, entityClass, entityType, Serialisers.NO_ANNOTATION, mediaType,
                             Serialisers.EMPTY_MULTI_MAP, writerOutput);
-                    writerOutput.writeTo(os);
                     wrote = true;
                 }
 

